src/app-layer-htp-range.c

```diff
@@ -221,110 +243,139 @@ void *HttpRangeContainerUrlGet(const uint8_t *key, size_t keylen, struct timeval
+        uint64_t end, uint64_t total, const StreamingBufferConfig *sbcfg, const uint8_t *name,
+        uint16_t name_len, uint16_t flags)
 {
-    SCMutexLock(&c->mutex);
+    DEBUG_VALIDATE_BUG_ON(c->files == NULL);
+
     if (c->files->tail == NULL) {
         if (FileOpenFileWithId(c->files, sbcfg, 0, name, name_len, NULL, 0, flags) != 0) {
             SCLogDebug("open file for range failed");
-            THashDecrUsecnt(c->hdata);
-            SCMutexUnlock(&c->mutex);
             return NULL;
         }
     }
     HttpRangeContainerBlock *curf = SCCalloc(1, sizeof(HttpRangeContainerBlock));
     if (curf == NULL) {
-        THashDecrUsecnt(c->hdata);
-        SCMutexUnlock(&c->mutex);
+        c->error = true;
         return NULL;
     }
     if (total > c->totalsize) {
         // TODOask add checks about totalsize remaining the same
         c->totalsize = total;
     }
-    uint64_t buflen = end - start + 1;
+    const uint64_t buflen = end - start + 1;
     if (start == c->files->tail->size && !c->appending) {
         // easy case : append to current file
         curf->container = c;
         c->appending = true;
-        SCMutexUnlock(&c->mutex);
         return curf;
     } else if (start < c->files->tail->size && c->files->tail->size - start >= buflen) {
         // only overlap
-        THashDecrUsecnt(c->hdata);
         // redundant to be explicit that this block is independent
-        curf->container = NULL;
         curf->toskip = buflen;
-        SCMutexUnlock(&c->mutex);
         return curf;
     } else if (start < c->files->tail->size && c->files->tail->size - start < buflen &&
                !c->appending) {
         // skip first overlap, then append
         curf->toskip = c->files->tail->size - start;
         c->appending = true;
         curf->container = c;
-        SCMutexUnlock(&c->mutex);
         return curf;
     }
-    // else {
     // block/range to be inserted in ordered linked list
     if (!(THASH_CHECK_MEMCAP(ContainerUrlRangeList.ht, buflen))) {
-        // TODOask release memory for other ranges cf RangeContainerFree(c);
         // skips this range
         curf->toskip = buflen;
-        curf->container = NULL;
-        THashDecrUsecnt(c->hdata);
-        SCMutexUnlock(&c->mutex);
         return curf;
     }
     curf->container = c;
-    (void)SC_ATOMIC_ADD(ContainerUrlRangeList.ht->memuse, buflen);
+
     HttpRangeContainerBuffer *range = SCCalloc(1, sizeof(HttpRangeContainerBuffer));
-    BUG_ON(range == NULL);
+    if (range == NULL) {
+        c->error = true;
+        SCFree(curf);
+        return NULL;
+    }
+
+    (void)SC_ATOMIC_ADD(ContainerUrlRangeList.ht->memuse, buflen);
     range->buffer = SCMalloc(buflen);
-    BUG_ON(range->buffer == NULL);
+    if (range->buffer == NULL) {
+        c->error = true;
+        SCFree(curf);
+        SCFree(range);
+        (void)SC_ATOMIC_SUB(ContainerUrlRangeList.ht->memuse, buflen);
```

Comment:
We could avoid this by having `SC_ATOMIC_ADD` a bit later

Reply:
Yeah, but the idea of the memcap updates it that we keep it as close to the actual allocation as possible to keep the accounting as accurate as we can. The error path should be very rare anyway.